File: openresty/lua/kvcache/hashing.lua
-- 前缀哈希(字符串版)。
-- 对应设计文档第 5.2 节。Lua 侧不做真实 BPE 分词(成本高),
-- 而是把前缀 messages 序列化成稳定字符串后做 SHA256,取前 16 个十六进制字符。
-- 稳定性/抗碰撞:用长度前缀 + 分隔符编码,避免 "ab"+"c" == "a"+"bc"。
local resty_sha256 = require "resty.sha256"
local resty_str = require "resty.string"   -- to_hex

local M = {}

-- 对任意字符串计算 SHA256,返回 16 位十六进制前缀。
-- (导出供 merkle.lua 复用)
M.sha256_hex16 = function(s)
    local sha = resty_sha256:new()
    sha:update(s)
    local bin = sha:final()
    return string.sub(resty_str.to_hex(bin), 1, 16)
end

local sha256_hex16 = M.sha256_hex16

-- 把单条 message 编码成稳定字符串(参与哈希)。
-- (导出供 merkle.lua 复用)
M.encode_message = function(msg)
    local role = msg.role or ""
    local content = msg.content
    if type(content) == "table" then
        -- 多模态/工具:用 cjson 的稳定序列化
        local cjson = require "cjson.safe"
        content = cjson.encode(content)
    else
        content = tostring(content or "")
    end
    -- #role 的长度前缀 + 分隔符,杜绝边界歧义
    return tostring(#role) .. ":" .. role .. "\x00" ..
           tostring(#content) .. ":" .. content .. "\x01"
end

local encode_message = M.encode_message

-- 对一段 messages 列表(前缀)计算哈希。
-- @param messages table: OpenAI messages 数组
-- @return string: 16 位十六进制哈希
function M.prefix_hash(messages)
    if not messages or #messages == 0 then
        return sha256_hex16("empty")
    end
    local parts = {}
    for i, msg in ipairs(messages) do
        parts[i] = encode_message(msg)
    end
    local blob = table.concat(parts, "")
    return sha256_hex16(blob)
end

-- 仅对前 prefix_len 条消息计算哈希。
function M.prefix_hash_n(messages, prefix_len)
    if prefix_len <= 0 then
        return M.prefix_hash(nil)
    end
    local sub = {}
    for i = 1, prefix_len do
        sub[i] = messages[i]
    end
    return M.prefix_hash(sub)
end

return M

File: tail/gateway/hashing.py
"""哈希算法(移植自 Lua hashing.lua,保证跨实现哈希一致)。

关键:encode_message 的字节格式必须与 Lua 版逐字节一致,
否则 Python gateway 与 OpenResty gateway 产出的 cache_key 会对不上。
"""

from __future__ import annotations

import hashlib
import json
from typing import List

HASH_HEX_LENGTH = 16


def sha256_hex16(s: str) -> str:
    """SHA256 取前 16 个十六进制字符(64 bit)。等价 Lua hashing.sha256_hex16。"""
    return hashlib.sha256(s.encode("utf-8")).hexdigest()[:HASH_HEX_LENGTH]


def _stable_content(content) -> str:
    """content 为 list/dict(多模态/工具)时,用固定字节序的 JSON 序列化。

    用 sort_keys + 紧凑分隔符,保证跨进程/跨语言稳定。
    Lua cjson 无 sort,但 Python 必须固定才能一致;此处是 Python 侧的规范。
    """
    if isinstance(content, (list, dict)):
        return json.dumps(content, separators=(",", ":"), sort_keys=True, ensure_ascii=False)
    return str(content if content is not None else "")


def encode_message(msg: dict) -> str:
    """单条 message 的稳定字符串编码(等价 Lua hashing.encode_message)。

    字节布局:<role_len>:<role>\\x00<content_len>:<content>\\x01
    长度用字节长度(UTF-8 编码后),防边界碰撞。
    """
    role = str(msg.get("role", ""))
    content = _stable_content(msg.get("content", ""))
    # 字节长度(Lua #role 是字节长度)
    rl = len(role.encode("utf-8"))
    cl = len(content.encode("utf-8"))
    return f"{rl}:{role}\x00{cl}:{content}\x01"


def prefix_hash(messages: List[dict]) -> str:
    """对 messages 数组算整体 hash(等价 Lua hashing.prefix_hash)。

    空列表返回对字面量 "empty" 的 hash(与 Lua 一致)。
    """
    if not messages:
        return sha256_hex16("empty")
    blob = "".join(encode_message(m) for m in messages)
    return sha256_hex16(blob)
